Use graphql api to pull  filter and pull sources

samshara opened the pull request:
- Fixes https://github.com/the-deep/client/issues/2140, #2067,  #2061
- Depends on https://github.com/the-deep/server/pull/960
- Depends on https://github.com/the-deep/server/pull/986

## Changes

* Use graphql to fetch filter options for sources filter
* Fetch list of sources using graphql api

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [ ] eslint issues
- [ ] typescript issues
- [ ] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [ ] permission checks
- [ ] translations

samshara committed:
Use graphql api to pull  filter and pull sources

samshara commented:
@tnagorra some typescript typing issues relating to the use of graphql and form library exist. Any idea on how we can resove it?

samshara committed:
Use graphql api for sources stats

samshara committed:
Add static entries filter to sources filter

tnagorra commented on app/views/Project/Tagging/Sources/index.tsx:
```diff
@@ -27,8 +26,8 @@ interface Props {
     const { project } = React.useContext(ProjectContext);
-    const activeProject = project ? +project.id : undefined;
-    const [sourcesFilters, setSourcesFilters] = useState<Filters>();
+    const activeProject = project ? project.id : undefined;
```
```suggestion
    const activeProject = project?.id;
```

tnagorra commented on app/views/Project/Tagging/Sources/utils.tsx:
```diff
@@ -55,3 +55,15 @@ export const getFiltersForRequest = (filters: FilterFormType | undefined) => {
     return requestFilters;
 };
+
+export function getValidDateRangeValues(dateRange?: { startDate: string, endDate: string}) {
```
Let's also add why this is useful in context with adding the timezone.

Also, do we need this on the new grapql endpoints?

samshara replied:
@tnagorra I used the logic from the previous filter. Even I don't know why we require a timezone offset here. Deep UI DateRangeInput does not support multiple name property and we require `createdAt_Gte` `createdAt_Lt`.

tnagorra replied:
Our DateInput component only supports `date` (without timezone offset)
Example: `2012-10-16`

But, the filters endDate and startDate accepts `datetime`.

When we are filtering with `datetime`, we should always send the timezone
information to the server.

tnagorra commented on app/views/Project/Tagging/Sources/SourcesStats/index.tsx:
```diff
@@ -111,18 +101,18 @@ function SourcesStats(props: Props) {
-                    title={_ts('sourcesStats', 'totalVerifiedEntries')}
+                    progress={((totalVerifiedEntries ?? 0) / (numberOfEntries ?? 0)) * 100}
+                    title="Entries Verified"
                     variant="complement1"
                     size="large"
                 />
                 <ProgressLine
-                    progress={(totalUnverifiedEntries / totalEntries) * 100}
-                    title={_ts('sourcesStats', 'totalUnverifiedEntries')}
+                    progress={((totalControlledEntries ?? 0) / (numberOfEntries ?? 0)) * 100}
```
We should prevent division by zero.
Let's create a small helper to calculate percentages or check if we have one in fujs

tnagorra commented on app/views/Project/Tagging/Sources/SourcesStats/index.tsx:
```diff
@@ -1,78 +1,64 @@
+        numberOfLeads,
+        numberOfLeadsTaggedAndControlled,
+        numberOfLeadsTagged,
+    } = sourcesStats?.project?.stats ?? {};
```
In TS, lets' not destrucutre on empty object

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/EntryFilter/index.tsx:
```diff
@@ -0,0 +1,134 @@
+const optionLabelSelector = (d: ControlStatusOption) => d.value;
+const optionKeySelector = (d: ControlStatusOption) => d.key;
+
+const controlledStatusOptions: ControlStatusOption[] = [
+    {
+        key: 'true',
+        value: 'Controlled',
+    },
+    {
+        key: 'false',
+        value: 'Not controlled',
+    },
+];
```
Does the server support true and false as string? Should these not be boolean?
P.S. deep-ui inputs don't support booleans as key

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -24,61 +24,232 @@ import {
 import styles from './styles.css';
 
-// FIXME: Created at and published on are date ranges and not date inputs
+export type SourcesFilterFields = PurgeNull<Pick<EnumFix<ProjectSourcesQueryVariables,
```
Why use Pick? We should not need to do this much of type manipulation

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -24,61 +24,232 @@ import {
+                description
+            }
+        }
+        project(id: $projectId) {
+            members {
+                id
+                displayName
+                firstName
+                lastName
+            }
+        }
```
members should not be pulled eagerly. These are later pulled using lazy select inputs

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -24,61 +24,232 @@ import {
+                lastName
+            }
+        }
+        organizationTypes {
+            results {
+                id
+                title
+            }
+        }
```
Is the no. of organization types limited?

tnagorra replied:
We should use OrganizationMultiSelectInput/OrganizationSelectInput instead

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -24,61 +24,232 @@ import {
+    }
+`;
+
+const userLabelSelector = (
+    d: NonNullable<SourceFilterOptionsQuery['project']>['members'][number],
+) => d.displayName ?? `${d.firstName} ${d.lastName}`;
+const userKeySelector = (
+    d: NonNullable<SourceFilterOptionsQuery['project']>['members'][number],
+) => d.id;
+const organizationTypeKeySelector = (
+    d: NonNullable<NonNullable<NonNullable<SourceFilterOptionsQuery['organizationTypes']>>['results']>[number],
+) => d.id;
+const organizationTypeLabelSelector = (
+    d: NonNullable<NonNullable<NonNullable<SourceFilterOptionsQuery['organizationTypes']>>['results']>[number],
+) => d.title;
```
Let's define types for Project and Organization to reuse them

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -24,61 +24,232 @@ import {
+    d: NonNullable<NonNullable<NonNullable<SourceFilterOptionsQuery['organizationTypes']>>['results']>[number],
+) => d.title;
+
+function getProjectSourcesQueryVariables(
```
It would be a lot better if we didn't need to do this transformation.

Can we generalize the actions we do there?

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -115,42 +271,73 @@ function SourcesFilter(props: Props) {
+            },
+        },
+    );
+    console.warn('loading', loading, sourceFilterOptionsError);
```
Let's not leave consle.warn on render

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -28,37 +29,35 @@ import {
 import styles from './styles.css';
 
-const leadsKeySelector: (d: Lead) => number = (d) => d.id;
+const sourcesKeySelector: (d: Lead) => string = (d) => d.id;
```
```suggestion
const sourcesKeySelector = (d: Lead) => d.id;
```

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -70,11 +69,94 @@ const defaultSorting = {
+        $search: String,
+        $statuses: [LeadStatusEnum!],
+        $entriesFilterData: LeadEntriesFilterData,
+        ) {
```
```suggestion
    ) {
```

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -70,11 +69,94 @@ const defaultSorting = {
+                    title
+                    publishedOn
+                    priority
+                    createdBy {
+                        displayName
+                    }
+                    project {
+                        id
+                    }
+                    authors {
+                        title
+                    }
+                    assignee {
+                        displayName
+                    }
```
```suggestion
                    createdBy {
                        id
                        displayName
                    }
                    project {
                        id
                    }
                    authors {
                        id
                        title
                    }
                    assignee {
                        id
                        displayName
                    }
```

Always get `id` field so that apollo client can properly update the normalized cache

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -83,13 +165,10 @@ function SourcesTable(props: Props) {
         refreshTimestamp,
     } = props;
 
+    console.warn('refres', refreshTimestamp);
```
Let's remove console.log on render

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -98,86 +177,74 @@ function SourcesTable(props: Props) {
-        getLeads();
-    }, [refreshTimestamp, getLeads]);
+    const sourcesResponse = projectSourcesResponse?.project?.leads;
+    const sources = sourcesResponse?.results as Lead[] | undefined | null;
```
We shouldn't need to cast this value

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -194,44 +261,44 @@ function SourcesTable(props: Props) {
-    const handleDelete = useCallback((leadId: number) => {
-        deleteLead(leadId);
-    }, [deleteLead]);
+    const handleDelete = useCallback((leadId: string) => {
+        console.warn('leadid', leadId);
+    }, []);
```
Let's add a 'TODO' or 'FIXME' comment

samshara committed:
Use calcPercent function to calculate percentage

samshara committed:
Add dynamic entry filters (widget filters) to sources filter

samshara committed:
Add GeoMultiSelectInput and use it in FrameworkFilterItem component

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -70,26 +69,108 @@ const defaultSorting = {
+                    authors {
+                        id
+                        title
+                    }
```
We should also fetch mergedAs for authors

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -98,86 +179,74 @@ function SourcesTable(props: Props) {
+                page: activePage,
+                pageSize: maxItemsPerPage,
+                ordering,
+                ...filters as Omit<ProjectSourcesQueryVariables, 'projectId' | 'page' | 'pageSize' | 'ordering'>,
```
Any reason we need to cast this?
Also, it's better to move this de-structuring at the start

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -70,26 +69,108 @@ const defaultSorting = {
-        offset: (activePage - 1) * maxItemsPerPage,
-        limit: maxItemsPerPage,
-    }), [activePage]);
+    const [selectedSources, setSelectedSources] = useState<Lead[]>([]);
```
We should not rename "lead" to "source".
The rename is only applicable for strings that users can see.

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/EntryFilter/FrameworkFilterItem/index.tsx:
```diff
@@ -0,0 +1,283 @@
+    switch (filter.widgetType) {
+        case 'DATE': {
+            return (
+                <>
```
We should hide/un-hide these filters if the filters are selected when "show all filters" is not enabled.

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -24,62 +26,284 @@ import {
+            }
+        }
+        project(id: $projectId) {
+            members {
+                id
+                displayName
+                firstName
+                lastName
+            }
```
We should use UserMultiSelectInput/UserSelectInput instead

samshara replied:
@tnagorra OrganizationMultiSelectInput/OrganizationSelectInput is for organizations. we are using organization types here.

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -24,62 +26,284 @@ import {
+            ...entriesFilters
+        } = entriesFilterData;
+
+        const controlled = controlledRaw === 'true';
```
Instead of adding this transformation here.
Let's create a input that supports boolean keys

samshara replied:
@tnagorra let's create another issue for it. to support boolean keys we need to make a lot of changes to the deep-ui.

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -24,62 +26,284 @@ import {
+    } = filters;
+
+    const createdAt = getValidDateRangeValues(createdAtRaw);
+    const publishedOn = getValidDateRangeValues(publishedOnRaw);
```
We should not transform publishedOn.
@thenav56 @AdityaKhatri we are storing publishedOn as date right?

thenav56 replied:
Yes published on is a date field.

samshara committed:
Create separate component for project member selection

** Show filters that have values

samshara committed:
Add BooleanInput component

Use BooleanInput component in control status filter in sources page

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/EntryFilter/index.tsx:
```diff
@@ -0,0 +1,202 @@
+
+const filterKeySelector = (d: AnalysisFrameworkFilterType) => d.key;
+
+const controlledStatusOptions: Option[] = [
+    {
+        key: 'true',
+        value: 'Controlled',
+    },
+    {
+        key: 'false',
+        value: 'Not controlled',
+    },
+];
```
Shouldn't we use BooleanInput

samshara replied:
We are using BooleanInput. The options are sent from the parent component

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -24,62 +26,274 @@ import {
+    | 'authoringOrganizationTypes'
+    | 'customFilters'
+>> & {
+    createdAt?: {
+        startDate: string;
+        endDate: string;
+    }
+    publishedOn?: {
+        startDate: string;
+        endDate: string;
+    }
```
We shouldn't need these object values for createdAt

samshara replied:
@tnagorra right now we are using DateRangeInput and I believe we require it this way.

tnagorra replied:
We have already created simplified date range input.

samshara replied:
@tnagorra  you mean https://github.com/the-deep/deep-ui/tree/feature/date-range-dual-input ? This does not work as expected right now. it updates the parent form object instead of the property itself. for example 
expected value after onChange
`
{ 
    startDate: '2021-1-11',
    endDate: '2021-2-22', 
 }` 
 actual output
 `{ startDate: '2021-1-12' } { endDate: '2021-2-22' } `
 `

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -116,43 +315,72 @@ function SourcesFilter(props: Props) {
+        ?.sourcePriorityOptions?.enumValues;
+    const confidentialityOptions = sourceFilterOptions
+        ?.sourceConfidentialityOptions?.enumValues;
+    const assigneesOptions = sourceFilterOptions
+        ?.project?.members;
```
We shouldn't pull project members directly

tnagorra commented on app/views/Project/Tagging/Sources/SourcesTable/index.tsx:
```diff
@@ -342,38 +414,43 @@ function SourcesTable(props: Props) {
+                'authors',
                 _ts('sourcesTable', 'authors'),
-                (item) => item?.authorsDetail.map((v) => v.title).join(', '),
+                (item) => item.authors?.map((v: Pick<OrganizationType, 'title' | 'mergedAs'>) => {
+                    if (v.mergedAs) {
+                        return v.mergedAs.title;
+                    }
+                    return v.title;
+                }).join(', '),
```
We already have a organization title selector in OrganizationSelectInput, let's use that

samshara committed:
Use DateRangeDualInput for date range filter

tnagorra committed:
Update usage of project members

- Add search on project members multiselect
- Fix typing issues

tnagorra committed:
Simplify entry filter transforms

- Create and use NumberButString component
- Create and use DateRnageDualInput component
- Fix typings issues

tnagorra committed:
Change default scalar type to "unknown"

- Use DateRangeInput with custom change handler for date filter

samshara committed:
Inject types for analysis framework filter options

tnagorra commented on app/views/Project/Tagging/Sources/utils.tsx:
```diff
@@ -55,3 +58,19 @@ export const getFiltersForRequest = (filters: FilterFormType | undefined) => {
     });
     return requestFilters;
 };
+
```
Let's check if we are using this function

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/types.ts:
```diff
@@ -0,0 +1,118 @@
+    label: string;
+}
+
+interface BaseProperties<T> {
+    type: string;
+    options?: T[];
+}
```
Just found out yesterday that properties can be undefined, so we should handle cases where properties are undefined and also define their types as nullable

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -24,62 +26,241 @@ import {
+    | 'entryTypes'
+>>;
+
+export type SourceFilterOptions = DeepReplace<SourceFilterOptionsQuery, Pick<AnalysisFrameworkFilterType, 'filterType' | 'key' | 'properties' | 'title' | 'widgetType'>, FrameworkFilterType>;
```
Any reason we need to use Pick?

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/index.tsx:
```diff
@@ -24,62 +26,241 @@ import {
+type FrameworkFilterFields = ReturnType<FrameworkFilterSchema['fields']>;
+const frameworkFilterSchema: FrameworkFilterSchema = {
+    fields: (): FrameworkFilterFields => ({
+        filterKey: [],
+        value: [],
+        valueList: [],
```
We should also add other values like valueGte, valueLte, etc.

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/EntryFilter/index.tsx:
```diff
@@ -0,0 +1,210 @@
+                )}
+                name="createdBy"
+                projectId={projectId}
+                value={value?.createdBy} // FIXME: fix type issue from server
```
Do we still have a typing issue here?

samshara replied:
yes we do

samshara replied:
yes we do

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/EntryFilter/FrameworkFilterItem/index.tsx:
```diff
@@ -0,0 +1,389 @@
+    switch (filter.widgetType) {
+        case 'DATE': {
+            return (
+                <DateInput
```
We may need to use ranged input for date, time and number inputs
@AdityaKhatri @thenav56 any thoughts?

thenav56 replied:
Yes.

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/EntryFilter/FrameworkFilterItem/index.tsx:
```diff
@@ -0,0 +1,389 @@
+                    options={geoAreaOptions}
+                    onOptionsChange={setGeoAreaOptions}
+                    disabled={disabled}
+                    placeholder={title}
```
We should add option to include sub-levels on geo-multiselect input

samshara replied:
added it but not testable right now as no GeoWidget in framework

tnagorra commented on app/views/Project/Tagging/Sources/SourcesFilter/EntryFilter/FrameworkFilterItem/index.tsx:
```diff
@@ -0,0 +1,389 @@
+                />
+            );
+        }
+        case 'ORGANIGRAM': { // FIXME options sent by the server is gibberish (null values)
```
@thenav56 we may have changed the data structure for organigram so the filterable data may not be correct for organigram.

thenav56 replied:
Current structure used in backend (from new branch). https://github.com/the-deep/server/blob/ebe7e51b91f08832ee17d54a0a5b5d12c0c85e09/apps/analysis_framework/widgets/organigram_widget.py#L3-L11.

Let me know if this still persists.

tnagorra replied:
Let me quicky verify if this is the case for me as well

samshara committed:
Use range values for number, date, time widgets

samshara committed:
Use range values for number, date, time widgets

tnagorra committed:
Fix problem with organigram widget

- Move "swap" button inside input
- Move "convertDateToIsoDateTime" to utils to avoid circular reference

tnagorra committed:
Set random string on organigram top node

AdityaKhatri commented on app/components/DateRangeDualInput/index.tsx:
```diff
@@ -0,0 +1,90 @@
+import React, { useCallback, useMemo } from 'react';
```
This is in deep-ui lets use it from here

samshara committed:
Use entry filters to filter entries under sources table

AdityaKhatri committed:
Fix empty entries filter
